streaming/storage/deque.go

```diff
@@ -0,0 +1,392 @@
+package storage
+
+import (
+	"github.com/cube2222/octosql"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+/*
+	Name clarification:
+	In a deque 1, 2, 3, 4 the Front is the 1 and the Back is the 4. This way
+	a Push/PopFront modify the front attributes, and Push/PopBack modify the back attributes.
+*/
+const (
+	defaultFirstFreeFrontSpot = 0
+	defaultFirstFreeBackSpot  = 1
+)
+
+var (
+	dequeValueKeyPrefix     = []byte("value")
+	dequeFirstFreeFrontSpot = []byte("first")
+	dequeFirstFreeBackSpot  = []byte("last")
+	defaultValues           = map[string]int{
+		string(dequeFirstFreeFrontSpot): defaultFirstFreeFrontSpot,
+		string(dequeFirstFreeBackSpot):  defaultFirstFreeBackSpot,
+	}
+)
+
+type Deque struct {
+	tx StateTransaction
+
+	initializedFront bool
+	initializedBack  bool
+
+	firstFreeFrontSpot int
+	firstFreeBackSpot  int
+}
+
+type DequeIterator struct {
+	it Iterator
+}
+
+func NewDeque(tx StateTransaction) *Deque {
+	return &Deque{
+		tx:                 tx,
+		initializedFront:   false,
+		initializedBack:    false,
+		firstFreeFrontSpot: defaultFirstFreeFrontSpot,
+		firstFreeBackSpot:  defaultFirstFreeBackSpot,
+	}
+}
+
+func NewDequeIterator(it Iterator) *DequeIterator {
+	return &DequeIterator{
+		it: it,
+	}
+}
+
+//Adds new element to the front of the queue
+func (dq *Deque) PushFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushFront")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return dq.appendBytes(data, dq.firstFreeFrontSpot, dq.decreaseFirstFreeFrontSpot)
+}
+
+//Adds new element to the back of the queue
```

Comment:
Your editor probably whines about this. Add a space after the beginning of a comment.

Reply:
It actually doesn't, but I changed it :)